### PR TITLE
Adjust EXP storage to uint64

### PR DESCRIPTION
**Pull request by cydh**
* **Addressed Issue(s)**: 
* **Server Mode**: Both
* **Description of Pull Request**: 
  * Changed EXP limit from uint32 to uint64 to support 185/65 updates
  * Changed EXP tables for 185/65 updates for #3342 maybe~

# PS
Just someone try to confirm if the EXP correct or not, [before they change the table!!](http://ro.gnjoy.com/guide/ragstart/play3.asp)
If someone say wrong without correction, I will revert the EXP changes!!
Forgot the job_db2.txt

### CODE REVIEW

**Comment by attackjom** [comment]
Very good job

**Comment by JohnnyPlayy** [comment]
It was very good😍
image
Red Aura works great on client 2018-06-21aRagexeRE

**Comment by stevenahoy** [comment]
im using 20180621a client how to get lvl 185 and using last rathena? are you edit manualy in database ? because my base lvl still in 175/60

**Comment by attackjom** [comment]
Can we skip these to 200/70 table

**Comment by aleos89** [comment]
It is the plan, but currently the table is a giant image from the release notes.

**Comment by cahya1992** [comment]
> It was very good😍
> image
> Red Aura works great on client 2018-06-21aRagexeRE

hi, how did you do to make red aura my aura still yellow

**Comment by RagnaWay** [comment]
> Can we skip these to 200/70 table

I agree, you should add 200/70 according to kRO

**Comment by cydh** [comment]
can't you fcking read the issue above?

**Comment by reunite-ro** [comment]
Since 3rd job 16th anniversary patch has been added, this pull should be good to be merged too right ?

**Comment by Lemongrass3110** [comment]
> Since 3rd job 16th anniversary patch has been added, this pull should be good to be merged too right ?

Not yet.

**Comment by RagnaWay** [comment]
the kRO website has already placed base **200** and job **70**. could update. 
http://ro.gnjoy.com/guide/ragstart/play3.asp

**Comment by RagnaWay** [comment]
update please

**Comment by admkakaroto** [comment]
This needs to be upgraded to level 200.

**Comment by admkakaroto** [comment]
Level 200 exp table: https://ro.gnjoy.com/news/update/View.asp?seq=242

**Comment by ecdarreola** [comment]
From the service_korea lua files the max job level for doram is still at 60. (latest data.grf -kro)

image

**Comment by aleos89** [comment]
Maybe they forgot to update the LUA? The table in @admkakaroto's link shows Doram going from 65 -> 70.

**Comment by ecdarreola** [comment]
> Maybe they forgot to update the LUA? The table in @admkakaroto's link shows Doram going from 65 -> 70.

Ah yeah maybe just lua forgotten.

File: src/char/int_guild.cpp [hunk 1]
```diff
@@ -377,7 +379,7 @@ struct guild * inter_guild_fromsql(int guild_id)
 	}
 	Sql_GetData(sql_handle,  5, &data, NULL); g->average_lv = atoi(data);
 	Sql_GetData(sql_handle,  6, &data, NULL); g->exp = strtoull(data, NULL, 10);
-	Sql_GetData(sql_handle,  7, &data, NULL); g->next_exp = (unsigned int)strtoul(data, NULL, 10);
+	Sql_GetData(sql_handle,  7, &data, NULL); g->next_exp = strtoull(data, NULL, 10);
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:382]:
```suggestion
	Sql_GetData(sql_handle,  7, &data, NULL); g->next_exp = strtoull(data, nullptr, 10);
```

File: src/char/int_guild.cpp [hunk 2]
```diff
@@ -1546,22 +1548,19 @@ int mapif_parse_GuildMemberInfoChange(int fd,int guild_id,uint32 account_id,uint
 		  }
 		case GMI_EXP:
 		{	// EXP
-			uint64 old_exp=g->member[i].exp;
-			g->member[i].exp=*((uint64 *)data);
+			t_exp old_exp=g->member[i].exp;
+			g->member[i].exp=*((t_exp *)data);
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:1552]: Please confirm this as working on x86

File: src/common/utils.cpp [hunk 3]
```diff
@@ -365,17 +365,38 @@ unsigned int get_percentage(const unsigned int A, const unsigned int B)
...
 	{
-		ShowError("get_percentage(): result percentage too high! (A=%u,B=%u,result=%g)\n", A, B, result);
+		ShowError("get_percentage: result percentage too high! (A=%u,B=%u,result=%g)\n", A, B, result);
 		return UINT_MAX;
 	}
 
 	return (unsigned int)floor(result);
 }
+
+uint32 get_percentage_exp(const uint64 A, const uint64 B)
```
Comment (Lemongrass3110) [inline src/common/utils.cpp:383]:
```suggestion
uint32 get_percentage_exp(const uint64 a, const uint64 b)
```

File: src/common/utils.cpp [hunk 4]
```diff
@@ -365,17 +365,38 @@ unsigned int get_percentage(const unsigned int A, const unsigned int B)
...
+{
+	double result;
+
+	if (B == 0)
```
Comment (Lemongrass3110) [inline src/common/utils.cpp:387]:
```suggestion
	if (b == 0)
```

File: src/common/utils.cpp [hunk 5]
```diff
@@ -365,17 +365,38 @@ unsigned int get_percentage(const unsigned int A, const unsigned int B)
...
+	{
+		ShowError("get_percentage_exp: divison by zero! (A=%" PRIu64 ",B=%" PRIu64 ")\n", A, B);
```
Comment (Lemongrass3110) [inline src/common/utils.cpp:389]:
```suggestion
		ShowError("get_percentage_exp: divison by zero! (a=%" PRIu64 ",b=%" PRIu64 ")\n", a, b);
```

File: src/common/utils.cpp [hunk 6]
```diff
@@ -365,17 +365,38 @@ unsigned int get_percentage(const unsigned int A, const unsigned int B)
...
+		return ~0U;
+	}
+
+	result = 100 * ((double)A / (double)B);
```
Comment (Lemongrass3110) [inline src/common/utils.cpp:393]:
```suggestion
	result = 100.0 * ((double)a / (double)b);
```

File: src/common/utils.cpp [hunk 7]
```diff
@@ -365,17 +365,38 @@ unsigned int get_percentage(const unsigned int A, const unsigned int B)
...
+
+	if (result > UINT32_MAX)
+	{
+		ShowError("get_percentage_exp: result percentage too high! (A=%" PRIu64 ",B=%" PRIu64 ",result=%g)\n", A, B, result);
```
Comment (Lemongrass3110) [inline src/common/utils.cpp:397]:
```suggestion
		ShowError("get_percentage_exp: result percentage too high! (a=%" PRIu64 ",b=%" PRIu64 ",result=%g)\n", a, b, result);
```

File: src/common/utils.hpp [hunk 8]
```diff
@@ -27,6 +27,7 @@ bool exists(const char* filename);
 
 /// calculates the value of A / B, in percent (rounded down)
 unsigned int get_percentage(const unsigned int A, const unsigned int B);
+uint32 get_percentage_exp(const uint64 A, const uint64 B);
```
Comment (Lemongrass3110) [inline src/common/utils.hpp:30]:
```suggestion
uint32 get_percentage_exp(const uint64 a, const uint64 b);
```

File: src/map/atcommand.cpp [hunk 9]
```diff
@@ -5313,11 +5313,11 @@ ACMD_FUNC(exp)
 	nullpo_retr(-1, sd);
 	memset(output, '\0', sizeof(output));
 
-	nextb = pc_nextbaseexp(sd);
+	nextb = (double)pc_nextbaseexp(sd);
 	if (nextb)
-		nextb = sd->status.base_exp*100.0/nextb;
+		nextb = nextb = sd->status.base_exp*100.0/nextb;
```
Comment (Lemongrass3110) [inline src/map/atcommand.cpp:5318]:
```suggestion
		nextb = sd->status.base_exp*100.0/nextb;
```

File: src/map/clif.cpp [hunk 10]
```diff
@@ -8525,8 +8535,8 @@ void clif_guild_basicinfo(struct map_session_data *sd) {
 	WFIFOL(fd,10)=g->connect_member;
 	WFIFOL(fd,14)=g->max_member;
 	WFIFOL(fd,18)=g->average_lv;
-	WFIFOL(fd,22)=(uint32)cap_value(g->exp,0,INT32_MAX);
-	WFIFOL(fd,26)=g->next_exp;
+	WFIFOL(fd,22)=(uint32)cap_value(g->exp, 0, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:8538]:
```suggestion
	WFIFOL(fd,22)=(uint32)cap_value(g->exp, 0, GEXP_MAX);
```

File: src/map/clif.cpp [hunk 11]
```diff
@@ -8525,8 +8535,8 @@ void clif_guild_basicinfo(struct map_session_data *sd) {
...
+	WFIFOL(fd,26)=(uint32)cap_value(g->next_exp, 0, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:8539]:
```suggestion
	WFIFOL(fd,26)=(uint32)cap_value(g->next_exp, 0, GEXP_MAX);
```

File: src/char/int_guild.cpp [hunk 12]
```diff
@@ -638,10 +640,10 @@ struct guild_castle* inter_guildcastle_fromsql(int castle_id)
 // Read exp_guild.txt
 bool exp_guild_parse_row(char* split[], int column, int current)
 {
-	unsigned int exp = (unsigned int)atol(split[0]);
+	t_exp exp = strtoull(split[0], nullptr, 10);
 
-	if (exp >= UINT_MAX) {
-		ShowError("exp_guild: Invalid exp %d at line %d\n", exp, current);
+	if (exp > EXP_MAX) {
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:645]:
```suggestion
	if (exp > GEXP_MAX) {
```

File: src/char/int_guild.cpp [hunk 13]
```diff
@@ -638,10 +640,10 @@ struct guild_castle* inter_guildcastle_fromsql(int castle_id)
...
+		ShowError("exp_guild: Invalid exp %" PRIu64 " at line %d, exceeds max of %" PRIu64 "\n", exp, current, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:646]:
```suggestion
		ShowError("exp_guild: Invalid exp %" PRIu64 " at line %d, exceeds max of %" PRIu64 "\n", exp, current, GEXP_MAX);
```

File: src/char/int_guild.cpp [hunk 14]
```diff
@@ -1546,22 +1548,19 @@ int mapif_parse_GuildMemberInfoChange(int fd,int guild_id,uint32 account_id,uint
...
 				// Compute gained exp
 				if (charserv_config.guild_exp_rate != 100)
 					exp = exp*(charserv_config.guild_exp_rate)/100;
 
 				// Update guild exp
-				if (exp > UINT64_MAX - g->exp)
-					g->exp = UINT64_MAX;
-				else
-					g->exp+=exp;
+				g->exp = util::safe_addition_cap(g->exp, exp, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:1563]:
```suggestion
				g->exp = util::safe_addition_cap(g->exp, exp, GEXP_MAX);
```

File: src/config/const.hpp [hunk 15]
```diff
@@ -53,6 +55,16 @@
...
+/**
+ * EXP definition type
+ */
+typedef uint64 t_exp;
+#if PACKETVER >= 20170830
+	const t_exp EXP_MAX = INT64_MAX;
+#else
+	const t_exp EXP_MAX = INT32_MAX;
+#endif
+
```
Comment (Lemongrass3110) [inline src/config/const.hpp:67]:
```suggestion

const t_exp GEXP_MAX = INT32_MAX;

```

File: src/map/guild.cpp [hunk 16]
```diff
@@ -402,10 +402,7 @@ int guild_payexp_timer_sub(DBKey key, DBData *data, va_list ap) {
 		return 0;
 	}
 
-	if (g->member[i].exp > UINT64_MAX - c->exp)
-		g->member[i].exp = UINT64_MAX;
-	else
-		g->member[i].exp+= c->exp;
+	g->member[i].exp = util::safe_addition_cap(g->member[i].exp, c->exp, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/map/guild.cpp:405]:
```suggestion
	g->member[i].exp = util::safe_addition_cap(g->member[i].exp, c->exp, GEXP_MAX);
```

File: src/map/guild.cpp [hunk 17]
```diff
@@ -1405,11 +1402,7 @@ unsigned int guild_payexp(struct map_session_data *sd,unsigned int exp) {
 	//Otherwise tax everything.
 
 	c = (struct guild_expcache *)db_data2ptr(guild_expcache_db->ensure(guild_expcache_db, db_i2key(sd->status.char_id), create_expcache, sd));
-
-	if (c->exp > UINT64_MAX - exp)
-		c->exp = UINT64_MAX;
-	else
-		c->exp += exp;
+	c->exp = util::safe_addition_cap(c->exp, exp, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/map/guild.cpp:1405]:
```suggestion
	c->exp = util::safe_addition_cap(c->exp, exp, GEXP_MAX);
```

File: src/map/guild.cpp [hunk 18]
```diff
@@ -1419,18 +1412,16 @@ unsigned int guild_payexp(struct map_session_data *sd,unsigned int exp) {
...
 
 	if (sd->status.guild_id == 0 || sd->guild == NULL)
 		return 0;
 
 	c = (struct guild_expcache *)db_data2ptr(guild_expcache_db->ensure(guild_expcache_db, db_i2key(sd->status.char_id), create_expcache, sd));
-	if (c->exp > UINT64_MAX - exp)
-		c->exp = UINT64_MAX;
-	else
-		c->exp += exp;
+	c->exp = util::safe_addition_cap(c->exp, exp, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/map/guild.cpp:1423]:
```suggestion
	c->exp = util::safe_addition_cap(c->exp, exp, GEXP_MAX);
```

File: src/map/pc.cpp [hunk 19]
```diff
@@ -8866,22 +8860,22 @@ bool pc_setparam(struct map_session_data *sd,int64 type,int64 val_tmp)
...
 		break;
 	case SP_BASEEXP:
-		{
-			val = cap_value(val, 0, INT_MAX);
-			if (val < sd->status.base_exp) // Lost
-				pc_lostexp(sd, sd->status.base_exp - val, 0);
-			else // Gained
-				pc_gainexp(sd, NULL, val - sd->status.base_exp, 0, 2);
-		}
+		val_tmp = cap_value(val_tmp, 0, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:8863]:
```suggestion
		val_tmp = cap_value(val_tmp, 0, pc_is_maxbaselv(sd) ? MAX_LEVEL_BASE_EXP : EXP_MAX);
```

File: src/map/pc.cpp [hunk 20]
```diff
@@ -8866,22 +8860,22 @@ bool pc_setparam(struct map_session_data *sd,int64 type,int64 val_tmp)
...
+		if (val_tmp < sd->status.base_exp) // Lost
+			pc_lostexp(sd, sd->status.base_exp - val_tmp, 0);
+		else // Gained
+			pc_gainexp(sd, NULL, val_tmp - sd->status.base_exp, 0, 2);
+		if (pc_is_maxbaselv(sd) && sd->status.base_exp > MAX_LEVEL_BASE_EXP)
+			sd->status.base_exp = MAX_LEVEL_BASE_EXP;
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:8869]:
```suggestion
```

File: src/map/pc.cpp [hunk 21]
```diff
@@ -8866,22 +8860,22 @@ bool pc_setparam(struct map_session_data *sd,int64 type,int64 val_tmp)
...
 		return true;
 	case SP_JOBEXP:
-		{
-			val = cap_value(val, 0, INT_MAX);
-			if (val < sd->status.job_exp) // Lost
-				pc_lostexp(sd, 0, sd->status.job_exp - val);
-			else // Gained
-				pc_gainexp(sd, NULL, 0, val - sd->status.job_exp, 2);
-		}
+		val_tmp = cap_value(val_tmp, 0, EXP_MAX);
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:8872]:
```suggestion
		val_tmp = cap_value(val_tmp, 0, pc_is_maxjoblv(sd) ? MAX_LEVEL_JOB_EXP : EXP_MAX);
```

File: src/map/pc.cpp [hunk 22]
```diff
@@ -8866,22 +8860,22 @@ bool pc_setparam(struct map_session_data *sd,int64 type,int64 val_tmp)
...
+		if (val_tmp < sd->status.job_exp) // Lost
+			pc_lostexp(sd, 0, sd->status.job_exp - val_tmp);
+		else // Gained
+			pc_gainexp(sd, NULL, 0, val_tmp - sd->status.job_exp, 2);
+		if (pc_is_maxjoblv(sd) && sd->status.job_exp > MAX_LEVEL_JOB_EXP)
+			sd->status.job_exp = MAX_LEVEL_JOB_EXP;
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:8878]:
```suggestion
```

File: src/map/pc.cpp [hunk 23]
```diff
@@ -12036,7 +12030,7 @@ static bool pc_readdb_job_exp(char* fields[], int columns, int current)
 
 	job_info[idx].max_level[type] = maxlvl;
 	for(i=0; i<maxlvl; i++)
-		job_info[idx].exp_table[type][i] = ((uint32) atoi(fields[3+i]));
+		job_info[idx].exp_table[type][i] = strtoull(fields[3+i], nullptr, 10);
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:12033]: Value caping to EXP_MAX? Including warning?

File: src/map/skill.cpp [hunk 24]
```diff
@@ -15803,7 +15803,7 @@ bool skill_check_condition_castbegin(struct map_session_data* sd, uint16 skill_i
 		case PR_REDEMPTIO:
 		case LG_INSPIRATION:
 			{
-				unsigned int exp, exp_needp = 0;
+				t_exp exp, exp_needp;
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:15806]:
```suggestion
				t_exp exp;
				uint32 exp_needp;
```

File: src/char/int_guild.cpp (hunk 2 above)
Comment (Lemongrass3110) [reply src/char/int_guild.cpp:1552]: Confirmed this as working.

File: src/map/pc.cpp (hunk 23 above)
Comment (Lemongrass3110) [reply src/map/pc.cpp]: Added in 7a4c334

**Comment by edwardlucian** [comment]
Captureds
Capturedss
Help plz

**Pull request merged**